crates/core/tedge_agent/src/tedge_to_te_converter/converter.rs

```diff
@@ -0,0 +1,153 @@
+use crate::tedge_to_te_converter::error::TedgetoTeConverterError;
+use clock::Timestamp;
+use log::error;
+use serde::Deserialize;
+use serde::Serialize;
+use serde_json::Value;
+use std::collections::HashMap;
+use std::convert::Infallible;
+use tedge_actors::Converter;
+use tedge_mqtt_ext::MqttMessage;
+use tedge_mqtt_ext::Topic;
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
+pub struct ThinEdgeAlarmData {
+    pub text: Option<String>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    #[serde(with = "time::serde::rfc3339::option")]
+    pub time: Option<Timestamp>,
+
+    #[serde(default)]
+    pub severity: String,
+
+    #[serde(flatten)]
+    pub alarm_data: HashMap<String, Value>,
+}
+
+pub struct TedgetoTeConverter {}
+
+impl Converter for TedgetoTeConverter {
+    type Input = MqttMessage;
+    type Output = MqttMessage;
+    type Error = Infallible;
+
+    fn convert(&mut self, input: &Self::Input) -> Result<Vec<Self::Output>, Self::Error> {
+        let messages_or_err = self.try_convert(input.clone());
+        Ok(self.wrap_errors(messages_or_err))
+    }
+}
+
+impl TedgetoTeConverter {
+    pub fn new() -> Self {
+        TedgetoTeConverter {}
+    }
+
+    fn try_convert(
+        &mut self,
+        mut message: MqttMessage,
+    ) -> Result<Vec<tedge_mqtt_ext::Message>, TedgetoTeConverterError> {
+        match message.topic.clone() {
+            topic if topic.name.starts_with("tedge/measurements") => {
+                Ok(self.convert_measurement(&mut message))
+            }
+            topic if topic.name.starts_with("tedge/events") => Ok(self.convert_event(&mut message)),
+            topic if topic.name.starts_with("tedge/alarms") => self.convert_alarm(&mut message),
+            topic if topic.name.starts_with("tedge/health") => {
+                Ok(self.convert_health_status_message(&mut message))
+            }
+            _ => Ok(vec![]),
+        }
+    }
+
+    // tedge/measurements -> te/device/main///m/
+    // tedge/measurements/child -> te/device/child///m/
+    fn convert_measurement(&mut self, message: &mut MqttMessage) -> Vec<tedge_mqtt_ext::Message> {
```

Comment:
I would update the message in place, avoiding to build a clone with `to_owned()`.

```suggestion
    fn convert_measurement(&mut self, mut message: MqttMessage) -> Vec<tedge_mqtt_ext::Message> {
```

Reply:
fixed